crates/bevy_animation/src/lib.rs

```diff
@@ -479,174 +485,217 @@ impl AnimationPlayer {
+                if let Ok((_, _, _, _, Some(parent_target))) = candidate_targets.get(parent.get()) {
+                    new_target = parent_target.clone();
+                }
+            }
+
+            // If we still didn't find an `AnimationTarget` for the parent, we could
+            // be an immediate child of an animation player. Handle that case.
+            if new_target.path.is_empty() {
+                if let Ok((_, _, _, name)) = players.get(parent.get()) {
+                    new_target = AnimationTarget {
+                        root: parent.get(),
+                        path: EntityPath::from(name.clone()),
+                    };
+                }
+            }
+        }
+
+        // If we still don't have an animation target after all that, we're part
+        // of a subtree that was once an animation target but got disconnected.
+        // Remove the `AnimationTarget` component entirely.
+        if new_target.path.is_empty() {
+            removed_targets.insert(entity);
         } else {
-            return true;
+            // Push our own name onto the path and record it.
+            new_target.path.parts.push(name.clone());
+            new_targets.insert(entity, new_target);
+        }
+
+        // Enqueue children.
+        if let Some(kids) = kids {
+            worklist.extend(kids);
         }
     }
+
+    // Finally, update the `AnimationTarget` components.
+    for (entity, animation_target) in new_targets.drain() {
+        commands.entity(entity).insert(animation_target);
+    }
+    for entity in removed_targets.drain() {
+        commands.entity(entity).remove::<AnimationTarget>();
+    }
 }
 
-/// System that will play all animations, using any entity with a [`AnimationPlayer`]
-/// and a [`Handle<AnimationClip>`] as an animation root
-#[allow(clippy::too_many_arguments)]
 pub fn animation_player(
     time: Res<Time>,
-    animations: Res<Assets<AnimationClip>>,
-    children: Query<&Children>,
-    names: Query<&Name>,
-    transforms: Query<&mut Transform>,
-    morphs: Query<&mut MorphWeights>,
-    parents: Query<(Has<AnimationPlayer>, Option<&Parent>)>,
-    mut animation_players: Query<(Entity, Option<&Parent>, &mut AnimationPlayer)>,
+    animation_clips: Res<Assets<AnimationClip>>,
+    mut players: Query<&mut AnimationPlayer>,
 ) {
-    animation_players
-        .par_iter_mut()
-        .for_each(|(root, maybe_parent, mut player)| {
-            update_transitions(&mut player, &time);
-            run_animation_player(
-                root,
-                player,
-                &time,
-                &animations,
-                &names,
-                &transforms,
-                &morphs,
-                maybe_parent,
-                &parents,
-                &children,
-            );
-        });
+    for mut player in players.iter_mut() {
+        update_transitions(&mut player, &time);
+
+        let paused = player.paused;
```

Comment:
Is `player.paused` affected by `update_transitions()`? If so, we may need to apply the animation once to ensure the final values are set before we stop animating. This is a bug we had in `bevy_tweening` where we stopped animating as soon as the animation finished, and so skipped applying the last frame. If so you can just move the `if paused { continue; }` at the top before `update_transitions()`.